File: scripts/batch-convert-tsx.py
#!/usr/bin/env python3
"""
Batch convert markdown files to TSX React components.
This script reads all .md files from content/lessons and content/videos
and generates corresponding .tsx component files.
"""

import os
import re
from pathlib import Path

def extract_title(content: str) -> str:
    """Extract title from markdown content (first ## heading)."""
    match = re.search(r'^##\s+(.+)$', content, re.MULTILINE)
    return match.group(1).strip() if match else "Untitled"

def slugify(filename: str) -> str:
    """Convert filename to slug."""
    # Remove leading numbers and .md extension
    slug = re.sub(r'^\d+-', '', filename)
    slug = re.sub(r'\.md$', '', slug)
    slug = slug.lower()
    slug = re.sub(r'[^a-z0-9]+', '-', slug)
    slug = re.sub(r'(^-|-$)', '', slug)
    return slug

def escape_jsx_string(s: str) -> str:
    """Escape string for use in JSX."""
    s = s.replace('\\', '\\\\')
    s = s.replace('{', '{{')
    s = s.replace('}', '}}')
    s = s.replace('`', '\\`')
    return s

def extract_quiz_questions(content: str):
    """Extract quiz questions from markdown content."""
    quiz_pattern = r'\*\*([^*]+\?)\*\*\s*\n((?:\|[^|]*\|[^|]*\|\n)*)'
    matches = re.findall(quiz_pattern, content)
    
    questions = []
    for question_text, options_text in matches:
        options = []
        for line in options_text.strip().split('\n'):
            if line.strip() and not re.match(r'^\|[\s:-]+\|[\s:-]+\|', line):
                cols = [c.strip() for c in line.split('|') if c.strip()]
                if len(cols) >= 2 and cols[0] == '':
                    options.append(cols[1])
                elif len(cols) == 1:
                    options.append(cols[0])
        
        if options:
            questions.append({
                'question': question_text.strip(),
                'options': options,
                'correctAnswer': 0  # Default to first option
            })
    
    return questions

def generate_component(md_file: str, component_type: str) -> str:
    """Generate a TSX component from a markdown file."""
    
    with open(md_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    title = extract_title(content)
    slug = slugify(os.path.basename(md_file))
    questions = extract_quiz_questions(content)
    
    # Remove quiz section from content
    content_without_quiz = re.sub(r'## Pop Quiz.*', '', content, flags=re.DOTALL).strip()
    
    # Extract main sections
    sections = []
    for match in re.finditer(r'^### (.+)$\n((?:[^\n]|\n(?!^###))*)', content_without_quiz, re.MULTILINE):
        section_title = match.group(1).strip()
        section_content = match.group(2).strip()
        sections.append((section_title, section_content))
    
    # Generate JSX for content
    content_jsx_lines = []
    for section_title, section_content in sections:
        content_jsx_lines.append(f'        <h2 className="text-2xl font-bold mt-6 mb-3">{escape_jsx_string(section_title)}</h2>')
        
        # Split content into paragraphs
        for para in section_content.split('\n\n'):
            para = para.strip()
            if para:
                escaped = escape_jsx_string(para)
                content_jsx_lines.append(f'        <p className="mb-4 leading-relaxed">{escaped}</p>')
    
    # Generate quiz section
    quiz_jsx = ""
    if questions:
        quiz_jsx = f"""
      <div className="mt-8">
        <Quiz 
          title="Pop Quiz" 
          questions={{{repr(questions)}}} 
          contentSlug="{slug}"
        />
      </div>"""
    
    component_code = f'''import {{ Quiz }} from "@/components/content/quiz";

export default function {title.replace(' ', '').replace('-', '')}Component() {{
  return (
    <div className="prose prose-lg max-w-none">
      <h1 className="text-3xl font-bold mb-4">{escape_jsx_string(title)}</h1>
{chr(10).join(content_jsx_lines)}
{quiz_jsx}
    </div>
  );
}}
'''
    
    return component_code

def convert_all_files():
    """Convert all markdown files to TSX components."""
    
    base_dir = Path('/vercel/share/v0-project')
    content_dirs = [
        (base_dir / 'content' / 'lessons', base_dir / 'components' / 'lessons', 'lessons'),
        (base_dir / 'content' / 'videos', base_dir / 'components' / 'videos', 'videos'),
    ]
    
    components_created = {'lessons': [], 'videos': []}
    
    for content_dir, component_dir, comp_type in content_dirs:
        if not content_dir.exists():
            print(f"Content directory not found: {content_dir}")
            continue
        
        component_dir.mkdir(parents=True, exist_ok=True)
        md_files = sorted(content_dir.glob('*.md'))
        
        print(f"\nProcessing {comp_type}...")
        print(f"Found {len(md_files)} markdown files")
        
        for idx, md_file in enumerate(md_files, 1):
            try:
                filename = md_file.name
                slug = slugify(filename)
                component_filename = f"{slug}.tsx"
                component_path = component_dir / component_filename
                
                # Skip if already exists (for manually converted files)
                if component_path.exists():
                    print(f"  [{idx}/{len(md_files)}] ⏭️  {component_filename} (already exists)")
                    components_created[comp_type].append((slug, filename))
                    continue
                
                # Generate and write component
                component_code = generate_component(str(md_file), comp_type)
                with open(component_path, 'w', encoding='utf-8') as f:
                    f.write(component_code)
                
                print(f"  [{idx}/{len(md_files)}] ✅ {component_filename}")
                components_created[comp_type].append((slug, filename))
                
            except Exception as e:
                print(f"  [{idx}/{len(md_files)}] ❌ {filename}: {str(e)[:50]}")
    
    # Print summary
    print(f"\n{'='*60}")
    print(f"Conversion Summary:")
    print(f"Lessons: {len(components_created['lessons'])} components")
    print(f"Videos: {len(components_created['videos'])} components")
    print(f"Total: {len(components_created['lessons']) + len(components_created['videos'])} components")
    print(f"{'='*60}")

if __name__ == '__main__':
    convert_all_files()

File: scripts/batch-convert-tsx.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '..');

const contentDir = path.join(projectRoot, 'content');
const lessonsContentDir = path.join(contentDir, 'lessons');
const videosContentDir = path.join(contentDir, 'videos');

const componentsDir = path.join(projectRoot, 'components');
const lessonsComponentDir = path.join(componentsDir, 'lessons');
const videosComponentDir = path.join(componentsDir, 'videos');

// Ensure component directories exist
if (!fs.existsSync(lessonsComponentDir)) {
  fs.mkdirSync(lessonsComponentDir, { recursive: true });
}
if (!fs.existsSync(videosComponentDir)) {
  fs.mkdirSync(videosComponentDir, { recursive: true });
}

function extractTitle(filename) {
  // Remove numeric prefix and .md extension: "0001-Title Here.md" -> "Title Here"
  return filename.replace(/^\d+-/, '').replace(/\.md$/, '');
}

function slugify(title) {
  return title
    .toLowerCase()
    .replace(/[^\w\s-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .trim();
}

function extractQuizzes(content) {
  const quizzes = [];
  const quizPattern = /##\s*Pop Quiz([\s\S]*?)(?=##\s|\Z)/g;
  let match;

  while ((match = quizPattern.exec(content)) !== null) {
    const quizContent = match[1];
    const questions = [];
    
    const questionPattern = /\*\*([^*]+)\*\*/g;
    let qMatch;
    const questionMatches = [];
    while ((qMatch = questionPattern.exec(quizContent)) !== null) {
      questionMatches.push({ question: qMatch[1].trim(), index: qMatch.index });
    }

    for (let i = 0; i < questionMatches.length; i++) {
      const { question, index } = questionMatches[i];
      const nextIndex = questionMatches[i + 1]?.index ?? quizContent.length;
      const section = quizContent.slice(index, nextIndex);

      const options = [];
      const lines = section.split('\n');
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed.startsWith('|')) continue;
        if (/^\|[\s:-]+\|[\s:-]+\|?/.test(trimmed)) continue;
        
        const cols = trimmed.split('|').map(c => c.trim()).filter((_, i, arr) => i > 0 && i < arr.length - 1);
        if (cols.length === 2 && cols[0] === '' && cols[1].length > 0) {
          options.push(cols[1]);
        }
      }

      if (options.length > 0) {
        questions.push({ question, options, correctAnswer: 0 });
      }
    }

    if (questions.length > 0) {
      quizzes.push(questions);
    }
  }

  return quizzes;
}

function generateTsxComponent(title, content, slug, quizzes) {
  // Strip Pop Quiz section from content
  const cleanContent = content.replace(/##\s*Pop Quiz[\s\S]*?(?=##\s|\Z)/g, '').trim();
  
  // Basic markdown to JSX conversion
  let jsxContent = cleanContent
    .split('\n')
    .map(line => {
      // Headers
      if (line.match(/^###\s/)) {
        return `<h3 className="text-xl font-bold mt-6 mb-3">${line.replace(/^###\s/, '').replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>')}</h3>`;
      }
      if (line.match(/^##\s/)) {
        return `<h2 className="text-2xl font-bold mt-8 mb-4">${line.replace(/^##\s/, '').replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>')}</h2>`;
      }
      if (line.match(/^#\s/)) {
        return `<h1 className="text-3xl font-bold mb-4">${line.replace(/^#\s/, '').replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>')}</h1>`;
      }
      // Bold
      line = line.replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>');
      // Italic
      line = line.replace(/\*(.+?)\*/g, '<em>$1</em>');
      // Inline code
      line = line.replace(/`(.+?)`/g, '<code className="bg-gray-100 px-2 py-1 rounded">$1</code>');
      
      if (line.trim().length > 0 && !line.startsWith('<')) {
        return `<p className="my-3 leading-relaxed">${line}</p>`;
      }
      return line;
    })
    .join('\n');

  const quizImports = quizzes.length > 0 ? `import { PopQuiz } from '@/components/pop-quiz';\n` : '';
  
  const quizComponents = quizzes.map((q, idx) => {
    return `        <PopQuiz 
          title="Pop Quiz"
          questions={quiz${idx}}
          contentSlug="${slug}"
        />`;
  }).join('\n');

  const quizData = quizzes.map((q, idx) => {
    return `const quiz${idx} = ${JSON.stringify(q)};`;
  }).join('\n\n');

  return `'use client';

${quizImports}
export default function ${title.split(' ').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join('')}() {
  ${quizData}

  return (
    <article className="prose prose-lg max-w-none">
      ${jsxContent}
      ${quizComponents}
    </article>
  );
}
`;
}

function convertFile(inputPath, outputDir, type) {
  try {
    const filename = path.basename(inputPath);
    const title = extractTitle(filename);
    const slug = slugify(title);
    const content = fs.readFileSync(inputPath, 'utf-8');
    const quizzes = extractQuizzes(content);
    
    const tsxContent = generateTsxComponent(title, content, slug, quizzes);
    const outputPath = path.join(outputDir, `${slug}.tsx`);
    
    fs.writeFileSync(outputPath, tsxContent);
    return { success: true, filename, slug, outputPath };
  } catch (error) {
    return { success: false, filename: path.basename(inputPath), error: error.message };
  }
}

function processDirectory(contentDirPath, componentDirPath, type) {
  console.log(`\n📂 Processing ${type}...`);
  
  if (!fs.existsSync(contentDirPath)) {
    console.error(`❌ Content directory not found: ${contentDirPath}`);
    return 0;
  }

  const files = fs.readdirSync(contentDirPath).filter(f => f.endsWith('.md'));
  console.log(`   Found ${files.length} markdown files`);

  let converted = 0;
  for (const file of files) {
    const inputPath = path.join(contentDirPath, file);
    const result = convertFile(inputPath, componentDirPath, type);
    
    if (result.success) {
      console.log(`   ✅ ${result.filename} → ${path.basename(result.outputPath)}`);
      converted++;
    } else {
      console.error(`   ❌ ${result.filename}: ${result.error}`);
    }
  }

  return converted;
}

console.log('🚀 Starting batch conversion of markdown to TSX...');
console.log(`Project root: ${projectRoot}`);

const lessonsConverted = processDirectory(lessonsContentDir, lessonsComponentDir, 'Lessons');
const videosConverted = processDirectory(videosContentDir, videosComponentDir, 'Videos');

console.log(`\n${'='.repeat(60)}`);
console.log(`Conversion Summary:`);
console.log(`Lessons: ${lessonsConverted} components`);
console.log(`Videos: ${videosConverted} components`);
console.log(`Total: ${lessonsConverted + videosConverted} components`);
console.log(`${'='.repeat(60)}\n`);

if (lessonsConverted + videosConverted > 0) {
  console.log('✅ Conversion complete! Update lib/content.ts to register new components.');
}
